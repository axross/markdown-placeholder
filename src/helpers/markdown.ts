import { camelCase, sentenceCase } from "change-case";
import { loremIpsum } from "lorem-ipsum";
import { type Paragraph, type PhrasingContent, type Text } from "mdast";
import prand from "pure-rand";

// eslint-disable-next-line no-unused-vars
type Rand = (min: number, max: number) => number;

const defaultSeed = 256;

function createRand({ seed }: { seed?: number } = {}): Rand {
  let rng = prand.xoroshiro128plus(seed ?? defaultSeed);

  function rand(min: number, max: number): number {
    const [value, nextRng] = prand.uniformIntDistribution(min, max, rng);

    rng = nextRng;

    return value;
  }

  return rand;
}

const floatRandGranularity = 4294967296;

function randSentence({ rand }: { rand: Rand }): string {
  const text = loremIpsum({
    count: 1,
    units: "sentence",
    suffix: "",
    random: () => rand(0, floatRandGranularity) / floatRandGranularity,
  });

  return `${sentenceCase(text)}.`;
}

const wrapperDefaultMaxLength = 3;

function wrapPartOfTextNodeWith(
  {
    text,
    start,
    end,
    rand,
  }: {
    text: Text;
    start?: number;
    end?: number;
    rand: Rand;
  },
  wrapper: (text: Text) => PhrasingContent,
): PhrasingContent[] {
  const words = text.value.split(" ");
  const resolvedStart = start ?? rand(0, words.length - 1);
  const resolvedEnd =
    end ??
    Math.min(resolvedStart + rand(1, wrapperDefaultMaxLength), words.length);
  const wordsBefore = words.slice(0, resolvedStart);
  const wordsToWrap = words.slice(resolvedStart, resolvedEnd);
  const wordsAfter = words.slice(resolvedEnd);
  const contents: PhrasingContent[] = [];

  if (wordsBefore.length > 0) {
    contents.push({
      type: "text",
      value: `${wordsBefore.join(" ")} `,
    });
  }

  contents.push(
    wrapper({
      type: "text",
      value: wordsToWrap.join(" "),
    }),
  );

  if (wordsAfter.length > 0) {
    contents.push({
      type: "text",
      value: ` ${wordsAfter.join(" ")}`,
    });
  }

  return contents;
}

const paragraphDefaultMinSentences = 1;
const paragraphDefaultMaxSentences = 8;
const paragraphChildOptionLength = 5;

function randParagraph({
  sentences,
  rand,
}: {
  sentences?: number;
  rand: Rand;
}): Paragraph {
  const resolvedSentences =
    sentences ??
    rand(paragraphDefaultMinSentences, paragraphDefaultMaxSentences);
  const contents: PhrasingContent[] = [];

  for (let index = 0; index < resolvedSentences; index += 1) {
    const sentence = randSentence({ rand });
    const text: Text = { type: "text", value: sentence };
    const chosen = rand(0, paragraphChildOptionLength);

    switch (chosen) {
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      case 1: {
        contents.push(
          ...wrapPartOfTextNodeWith({ text, rand }, (node) => ({
            type: "strong",
            children: [node],
          })),
        );

        break;
      }
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      case 2: {
        contents.push(
          ...wrapPartOfTextNodeWith({ text, rand }, (node) => ({
            type: "emphasis",
            children: [node],
          })),
        );

        break;
      }
      // eslint-disable-next-line @typescript-eslint/no-magic-numbers
      case 3: {
        contents.push(
          ...wrapPartOfTextNodeWith({ text, rand }, (node) => ({
            type: "inlineCode",
            value: camelCase(node.value),
          })),
        );

        break;
      }
      default: {
        contents.push(text);
      }
    }
  }

  return {
    type: "paragraph",
    children: contents,
  };
}

export { createRand, randParagraph };
