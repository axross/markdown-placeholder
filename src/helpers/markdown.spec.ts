import { describe, expect, it } from "vitest";
import { createRand, randText } from "./markdown";

describe("randText()", () => {
  it("returns a text node with random string value", () => {
    const rand = createRand();

    expect(randText({ rand })).toMatchInlineSnapshot(`
      {
        "type": "text",
        "value": "nulla aliqua magna sint aliqua amet dolore veniam",
      }
    `);
  });

  it("continuously returns text nodes with random string value when you call it more than once", () => {
    const rand = createRand();

    expect(randText({ rand })).toMatchInlineSnapshot(`
      {
        "type": "text",
        "value": "nulla aliqua magna sint aliqua amet dolore veniam",
      }
    `);
    expect(randText({ rand })).toMatchInlineSnapshot(`
      {
        "type": "text",
        "value": "ex veniam ipsum",
      }
    `);
    expect(randText({ rand })).toMatchInlineSnapshot(`
      {
        "type": "text",
        "value": "excepteur consectetur labore",
      }
    `);
  });
});
