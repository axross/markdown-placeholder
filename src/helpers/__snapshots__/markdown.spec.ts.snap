// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`randParagraph() > continuously returns paragraph nodes with random string value when you call it more than once 1`] = `
{
  "children": [
    {
      "type": "text",
      "value": "Excepteur consectetur labore dolor id ullamco nostrud in nulla aliqua magna sint aliqua ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "amet dolore.",
        },
      ],
      "type": "strong",
    },
    {
      "type": "text",
      "value": "Velit excepteur voluptate officia aliquip aliqua lorem anim nisi mollit nisi mollit est labore.",
    },
    {
      "type": "text",
      "value": "Laborum aliqua dolor dolor sunt non reprehenderit eu do laborum sit.",
    },
    {
      "type": "text",
      "value": "Proident incididunt nostrud lorem nulla irure laborum labore.",
    },
    {
      "type": "text",
      "value": "Eiusmod sint ipsum voluptate veniam labore pariatur ad proident.",
    },
    {
      "type": "text",
      "value": "Id culpa sunt lorem laborum ullamco consequat et anim do ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "sit",
        },
      ],
      "type": "strong",
    },
    {
      "type": "text",
      "value": " aliquip non culpa elit.",
    },
    {
      "type": "text",
      "value": "Irure sint esse laboris in est ipsum ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "non",
        },
      ],
      "type": "emphasis",
    },
    {
      "type": "text",
      "value": " commodo consequat.",
    },
    {
      "type": "text",
      "value": "Velit non in ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "et id amet",
        },
      ],
      "type": "emphasis",
    },
    {
      "type": "text",
      "value": " ea.",
    },
  ],
  "type": "paragraph",
}
`;

exports[`randParagraph() > continuously returns paragraph nodes with random string value when you call it more than once 2`] = `
{
  "children": [
    {
      "type": "text",
      "value": "Est in deserunt pariatur esse nisi ut culpa.",
    },
    {
      "type": "text",
      "value": "Duis et dolor ipsum sit laborum elit quis sit consequat.",
    },
    {
      "type": "text",
      "value": "Mollit laboris veniam nisi do cupidatat enim laboris adipisicing pariatur in cillum consequat.",
    },
    {
      "type": "text",
      "value": "Ex velit enim reprehenderit amet magna occaecat laboris.",
    },
    {
      "type": "text",
      "value": "Duis officia ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "lorem exercitation",
        },
      ],
      "type": "emphasis",
    },
    {
      "type": "text",
      "value": " adipisicing ad pariatur anim commodo eu officia.",
    },
    {
      "type": "text",
      "value": "Anim ex elit est cillum amet laborum consequat sit velit reprehenderit adipisicing laborum officia ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "amet.",
        },
      ],
      "type": "emphasis",
    },
  ],
  "type": "paragraph",
}
`;

exports[`randParagraph() > continuously returns paragraph nodes with random string value when you call it more than once 3`] = `
{
  "children": [
    {
      "type": "text",
      "value": "In nulla laboris occaecat tempor ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "id.",
        },
      ],
      "type": "emphasis",
    },
    {
      "type": "text",
      "value": "Dolore laborum nostrud nulla fugiat irure proident reprehenderit est anim et qui ad sunt ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "sunt.",
        },
      ],
      "type": "strong",
    },
    {
      "type": "text",
      "value": "Cupidatat ipsum reprehenderit minim ut anim laborum est proident reprehenderit est exercitation.",
    },
    {
      "type": "text",
      "value": "Nulla aliquip ut do qui laborum ullamco reprehenderit est do dolor esse ",
    },
    {
      "type": "inlineCode",
      "value": "doloreEst",
    },
    {
      "type": "text",
      "value": "Officia labore enim aute nostrud.",
    },
  ],
  "type": "paragraph",
}
`;

exports[`randParagraph() > returns a paragraph node with random string value 1`] = `
{
  "children": [
    {
      "type": "text",
      "value": "Excepteur consectetur labore dolor id ullamco nostrud in nulla aliqua magna sint aliqua ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "amet dolore.",
        },
      ],
      "type": "strong",
    },
    {
      "type": "text",
      "value": "Velit excepteur voluptate officia aliquip aliqua lorem anim nisi mollit nisi mollit est labore.",
    },
    {
      "type": "text",
      "value": "Laborum aliqua dolor dolor sunt non reprehenderit eu do laborum sit.",
    },
    {
      "type": "text",
      "value": "Proident incididunt nostrud lorem nulla irure laborum labore.",
    },
    {
      "type": "text",
      "value": "Eiusmod sint ipsum voluptate veniam labore pariatur ad proident.",
    },
    {
      "type": "text",
      "value": "Id culpa sunt lorem laborum ullamco consequat et anim do ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "sit",
        },
      ],
      "type": "strong",
    },
    {
      "type": "text",
      "value": " aliquip non culpa elit.",
    },
    {
      "type": "text",
      "value": "Irure sint esse laboris in est ipsum ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "non",
        },
      ],
      "type": "emphasis",
    },
    {
      "type": "text",
      "value": " commodo consequat.",
    },
    {
      "type": "text",
      "value": "Velit non in ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "et id amet",
        },
      ],
      "type": "emphasis",
    },
    {
      "type": "text",
      "value": " ea.",
    },
  ],
  "type": "paragraph",
}
`;
